### PR TITLE
OopenSSL 1.1.0 compatibility

### LINKED ISSUE #104: Doesn't build with openssl 1.1
yubico-piv-tool fails to build with the latest openssl release 1.1 (which has made some API changes), please see this Debian bug for a build log and pointers to documentation for the API changes:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=828616

### PR DESCRIPTION
**Pull request by Jakuje**
Update 2017-11-14: Resolves  #104. Tested with latest Yubikey 4 with hardware tests, provisioning new key and selfsigning certificates.

### CODE REVIEW

File: tool/yubico-piv-tool.c [hunk 1]
```diff
@@ -724,9 +724,7 @@ static bool request_certificate(ykpiv_state *state, enum enum_key_format key_for
   }
 
   memcpy(digest, oid, oid_len);
-  /* XXX: this should probably use X509_REQ_digest() but that's buggy */
-  if(!ASN1_item_digest(ASN1_ITEM_rptr(X509_REQ_INFO), md, req->req_info,
-              digest + oid_len, &digest_len)) {
+  if (!X509_REQ_digest(req, md, digest + oid_len, &digest_len)) {
```
Comment (klali) [inline tool/yubico-piv-tool.c]: Have you tested if this works? The problem with  X509_REQ_digest() is (used to be?) that it hashes the entire X509_REQ, ideally I guess there would be a X509_REQ_INFO_digest().

**Comment by klali** [comment]
The approach of doing this with a compat layer is probably the only way to go ahead. I haven't looked it over to see how many of those would need to be added.
There is no direct plan or schedule for this work.

File: tool/yubico-piv-tool.c (hunk 1 above)
Comment (Jakuje) [inline tool/yubico-piv-tool.c]: I didn't test this part, but I am not sure if there is an interface in the new OpenSSL to get the `req_info` from the structure. I will remove this chunk for now.
I planned to do all the changes, but some of the certificate handling was more complex than I was familiar with. Currently dumping what is available for others to have possibility to continue or I will finish that later.

**Comment by Jakuje** [comment]
Thank you for prompt comment. There are really only few more changes, but they are probably harder to address than this part in this commit.
If there is no plan yet from Yubico, I will be hopefully able to continue on that (or somebody else), since many distributions are moving to openssl 1.1.0

**Comment by Jakuje** [comment]
I added few more changes I was able to handle. The embedded tests work except the hardware ones (failing even before the change).

What is missing is working with `{X509,X509_REQ}_digest()` and tweaking of `{x509,req}->signature->flags` and `{req,x509}->{cert_info,sig_alg}->{parameter,algorithm}` on several places in the code.

I believe that is a good start, but I don't have enough internal knowledge to test further changes to make sure they will not break some use cases. I will build the package against old OpenSSL so far, but having this fixed in future would be very nice.

**Comment by Jakuje** [comment]
Finally my new Yubikey arrived so before setting it up for myself I finished with this PR and made even the HW tests pass with OpenSSL 1.1.0 and hopefully also with the old version (including the rebase to the current master).

**Comment by mrmekon** [comment]
@Jakuje I updated this to work with the recent refactoring: PR #131

**Pull request merged**